sql: logic test panics with --verbosity 3

```
$ make test PKG=./pkg/sql/logictest TESTS=Logic/distsql/join 'TESTFLAGS=-v --show-logs --show-sql --verbosity 3'
...
panic: unresolved names ought to be replaced before this point [recovered]
	panic: panic while executing "SELECT * FROM onecolumn JOIN twocolumn ON twocolumn.x = onecolumn.x AND onecolumn.x IN (SELECT x FROM twocolumn WHERE y >= 52)": unresolved names ought to be replaced before this point [recovered]
	panic: panic while executing "SELECT * FROM onecolumn JOIN twocolumn ON twocolumn.x = onecolumn.x AND onecolumn.x IN (SELECT x FROM twocolumn WHERE y >= 52)": unresolved names ought to be replaced before this point [recovered]
	panic: panic while executing "SELECT * FROM onecolumn JOIN twocolumn ON twocolumn.x = onecolumn.x AND onecolumn.x IN (SELECT x FROM twocolumn WHERE y >= 52)": unresolved names ought to be replaced before this point

goroutine 690 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc4200adb00, 0x2ab14ee3b640, 0xc420326630)
	/go2/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:180 +0xb1
panic(0x182f2a0, 0xc4202b3d60)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve.func1(0xc420cf8800, 0x2ab14ee3b640, 0xc420a170b0)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:360 +0xa1
panic(0x182f2a0, 0xc4202b3d60)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql.(*Session).maybeRecover(0xc4213d2000, 0x19bb989, 0x9, 0xc4212fd311, 0x7e)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/session.go:1165 +0x28d
panic(0x178eb80, 0xc42032b020)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql/parser.UnresolvedName.ResolvedType(0xc420320bb0, 0x1, 0x1, 0x19b34f4, 0x2)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/var_name.go:101 +0x64
github.com/cockroachdb/cockroach/pkg/sql/parser.(*UnresolvedName).ResolvedType(0xc4202afcc0, 0x19b34f4, 0x2)
	<autogenerated>:379 +0x5f
github.com/cockroachdb/cockroach/pkg/sql/parser.FormatNode(0xc4213f9a40, 0xc420cab020, 0x25e5760, 0xc4202afcc0)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/format.go:154 +0x313
github.com/cockroachdb/cockroach/pkg/sql/parser.SelectExpr.Format(0x25fa6e0, 0xc4202afcc0, 0x0, 0x0, 0xc4213f9a40, 0xc420cab020)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/select.go:140 +0x71
github.com/cockroachdb/cockroach/pkg/sql/parser.(*SelectExpr).Format(0xc4202b3ce0, 0xc4213f9a40, 0xc420cab020)
	<autogenerated>:399 +0x7c
github.com/cockroachdb/cockroach/pkg/sql/parser.formatNodeOrHideConstants(0xc4213f9a40, 0xc420cab020, 0x25e54a0, 0xc4202b3ce0)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/hide_constants.go:42 +0xc0
github.com/cockroachdb/cockroach/pkg/sql/parser.FormatNode(0xc4213f9a40, 0xc420cab020, 0x25e54a0, 0xc4202b3ce0)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/format.go:167 +0x67
github.com/cockroachdb/cockroach/pkg/sql/parser.SelectExprs.Format(0xc4202afd20, 0x1, 0x1, 0xc4213f9a40, 0xc420cab020)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/select.go:107 +0xdc
github.com/cockroachdb/cockroach/pkg/sql/parser.(*SelectExprs).Format(0xc4202b3cc0, 0xc4213f9a40, 0xc420cab020)
	<autogenerated>:400 +0x73
github.com/cockroachdb/cockroach/pkg/sql/parser.formatNodeOrHideConstants(0xc4213f9a40, 0xc420cab020, 0x25e54e0, 0xc4202b3cc0)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/hide_constants.go:42 +0xc0
github.com/cockroachdb/cockroach/pkg/sql/parser.FormatNode(0xc4213f9a40, 0xc420cab020, 0x25e54e0, 0xc4202b3cc0)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/format.go:167 +0x67
github.com/cockroachdb/cockroach/pkg/sql/parser.(*SelectClause).Format(0xc4213f7dc0, 0xc4213f9a40, 0xc420cab020)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/select.go:89 +0x140
github.com/cockroachdb/cockroach/pkg/sql/parser.formatNodeOrHideConstants(0xc4213f9a40, 0xc420cab020, 0x25e18a0, 0xc4213f7dc0)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/hide_constants.go:42 +0xc0
github.com/cockroachdb/cockroach/pkg/sql/parser.FormatNode(0xc4213f9a40, 0xc420cab020, 0x25e18a0, 0xc4213f7dc0)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/format.go:167 +0x67
github.com/cockroachdb/cockroach/pkg/sql/parser.(*Select).Format(0xc420caa810, 0xc4213f9a40, 0xc420cab020)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/select.go:50 +0x73
github.com/cockroachdb/cockroach/pkg/sql/parser.formatNodeOrHideConstants(0xc4213f9a40, 0xc420cab020, 0x25e1860, 0xc420caa810)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/hide_constants.go:42 +0xc0
github.com/cockroachdb/cockroach/pkg/sql/parser.FormatNode(0xc4213f9a40, 0xc420cab020, 0x25e1860, 0xc420caa810)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/format.go:167 +0x67
github.com/cockroachdb/cockroach/pkg/sql/parser.(*ParenSelect).Format(0xc4204a82a8, 0xc4213f9a40, 0xc420cab020)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/select.go:63 +0x61
github.com/cockroachdb/cockroach/pkg/sql/parser.formatNodeOrHideConstants(0xc4213f9a40, 0xc420cab020, 0x25e13e0, 0xc4204a82a8)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/hide_constants.go:42 +0xc0
github.com/cockroachdb/cockroach/pkg/sql/parser.FormatNode(0xc4213f9a40, 0xc420cab020, 0x25e13e0, 0xc4204a82a8)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/format.go:167 +0x67
github.com/cockroachdb/cockroach/pkg/sql/parser.(*Subquery).Format(0xc420320cb0, 0xc4213f9a40, 0xc420cab020)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/expr.go:730 +0x6f
github.com/cockroachdb/cockroach/pkg/sql/parser.formatNodeOrHideConstants(0xc4213f9a40, 0xc420cab020, 0x25e1ee0, 0xc420320cb0)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/hide_constants.go:42 +0xc0
github.com/cockroachdb/cockroach/pkg/sql/parser.FormatNode(0xc4213f9a40, 0xc420cab020, 0x25e1ee0, 0xc420320cb0)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/format.go:167 +0x67
github.com/cockroachdb/cockroach/pkg/sql.(*subquery).Format(0xc42023a640, 0xc4213f9a40, 0xc420cab020)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/subquery.go:82 +0x6f
github.com/cockroachdb/cockroach/pkg/sql/parser.formatNodeOrHideConstants(0xc4213f9a40, 0xc420cab020, 0x25df620, 0xc42023a640)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/hide_constants.go:42 +0xc0
github.com/cockroachdb/cockroach/pkg/sql/parser.FormatNode(0xc4213f9a40, 0xc420cab020, 0x25df620, 0xc42023a640)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/format.go:152 +0x2d9
github.com/cockroachdb/cockroach/pkg/sql/parser.exprFmtWithParen(0xc4213f9a40, 0xc420cab020, 0x25f8b20, 0xc42023a640)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/expr.go:109 +0x122
github.com/cockroachdb/cockroach/pkg/sql/parser.binExprFmtWithParenAndSubOp(0xc4213f9a40, 0xc420cab020, 0x25f9860, 0xc420caaed0, 0x0, 0x0, 0x19b3598, 0x2, 0x25f8b20, 0xc42023a640)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/expr.go:155 +0xd1
github.com/cockroachdb/cockroach/pkg/sql/parser.binExprFmtWithParen(0xc4213f9a40, 0xc420cab020, 0x25f9860, 0xc420caaed0, 0x19b3598, 0x2, 0x25f8b20, 0xc42023a640)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/expr.go:141 +0x93
github.com/cockroachdb/cockroach/pkg/sql/parser.(*ComparisonExpr).Format(0xc4213b8d00, 0xc4213f9a40, 0xc420cab020)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/expr.go:382 +0x131
github.com/cockroachdb/cockroach/pkg/sql/parser.formatNodeOrHideConstants(0xc4213f9a40, 0xc420cab020, 0x25e03a0, 0xc4213b8d00)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/hide_constants.go:42 +0xc0
github.com/cockroachdb/cockroach/pkg/sql/parser.FormatNode(0xc4213f9a40, 0xc420cab020, 0x25e03a0, 0xc4213b8d00)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/format.go:152 +0x2d9
github.com/cockroachdb/cockroach/pkg/sql/parser.AsStringWithFlags(0x25e03a0, 0xc4213b8d00, 0xc420cab020, 0x25e03a0, 0xc4213b8d00)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/parser/format.go:191 +0x62
github.com/cockroachdb/cockroach/pkg/sql.(*explainer).expr(0xc420402a80, 0x19b4a2e, 0x4, 0x19b704a, 0x6, 0xffffffffffffffff, 0x25f91e0, 0xc4213b8d00)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/explain_plan.go:189 +0x107
github.com/cockroachdb/cockroach/pkg/sql.(*explainer).(github.com/cockroachdb/cockroach/pkg/sql.expr)-fm(0x19b4a2e, 0x4, 0x19b704a, 0x6, 0xffffffffffffffff, 0x25f91e0, 0xc4213b8d00)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/explain_plan.go:171 +0x73
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).expr(0xc420e2dd40, 0x19b4a2e, 0x4, 0x19b704a, 0x6, 0xffffffffffffffff, 0x25f91e0, 0xc4213b8d00, 0x0, 0x0, ...)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:408 +0x1eb
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit(0xc420e2dd40, 0x25f89a0, 0xc42118a000)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:156 +0x19aa
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit(0xc420e2dd40, 0x25f88a0, 0xc4213d5a00)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:206 +0xaf6
github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit(0xc420e2dd40, 0x25f8960, 0xc421101100)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:170 +0x3ab2
github.com/cockroachdb/cockroach/pkg/sql.walkPlan(0x2ab14ee3b640, 0xc420caa960, 0x25f8960, 0xc421101100, 0xc42032a760, 0xc42032a770, 0xc42032a7f0, 0xc42032a800, 0x0, 0xc420caa960, ...)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:60 +0x18f
github.com/cockroachdb/cockroach/pkg/sql.planToString(0x2ab14ee3b640, 0xc420caa960, 0x25f8960, 0xc421101100, 0xc421101100, 0xc420321c59)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/explain_plan.go:164 +0x20b
github.com/cockroachdb/cockroach/pkg/sql.(*planner).makePlan(0xc4213d21d8, 0x2ab14ee3b640, 0xc420caa960, 0x25f9c60, 0xc420caa900, 0x0, 0x0, 0x0, 0xc4201ce680, 0x596e22d5, ...)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:199 +0x239
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmt(0xc42041c400, 0x25f9c60, 0xc420caa900, 0x0, 0x0, 0x0, 0xc4201ce680, 0xc4213d21d8, 0x0, 0xed1001900, ...)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1691 +0x11b
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmtInOpenTxn(0xc42041c400, 0xc4213d2000, 0x25f9c60, 0xc420caa900, 0x0, 0x0, 0x0, 0xc4201ce680, 0x0, 0x1, ...)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1436 +0x5ed
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execStmtsInCurrentTxn(0xc42041c400, 0xc4213d2000, 0xc420caa930, 0x1, 0x1, 0x0, 0x1, 0x0, 0x0, 0x0, ...)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:1059 +0x98f
github.com/cockroachdb/cockroach/pkg/sql.runTxnAttempt(0xc42041c400, 0xc4213d2000, 0xc420caa930, 0x1, 0x1, 0x0, 0x1, 0xc420320de8, 0xc420ca0001, 0x0, ...)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:968 +0xaf
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execParsed.func1(0x2ab14ee3b640, 0xc420caa960, 0xc420521340, 0xc420320de8, 0x0, 0x0)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:773 +0x24b
github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Exec(0xc420521340, 0x2ab14ee3b640, 0xc420caa960, 0x10101, 0xc4213b8300, 0x18f49a31, 0x0)
	/go2/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:675 +0xb6
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execParsed(0xc42041c400, 0xc4213d2000, 0xc420caa930, 0x1, 0x1, 0x0, 0x0, 0x4d, 0x0, 0x4d, ...)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:795 +0x48e
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).execRequest(0xc42041c400, 0xc4213d2000, 0xc4212fd311, 0x7e, 0x0, 0x0, 0x7e, 0xc4212fd311, 0x7e, 0xcef)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:674 +0x594
github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements(0xc42041c400, 0xc4213d2000, 0xc4212fd311, 0x7e, 0x0, 0x0, 0x0, 0x0, 0xc400000000)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/executor.go:550 +0x149
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc420cf8800, 0xc420cf8828, 0x19cb0fb, 0x15)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:519 +0xb2
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve(0xc420cf8800, 0x2ab14ee3b640, 0xc420a170b0, 0xc420294e60, 0x5400, 0xc420400310, 0x0, 0x0)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:440 +0x8bb
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*Server).ServeConn(0xc4204001c0, 0x2ab14ee3b680, 0xc420e9a600, 0x2600860, 0xc421085c00, 0x0, 0x0)
	/go2/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/server.go:421 +0x926
github.com/cockroachdb/cockroach/pkg/server.(*Server).Start.func8.1(0x2601ac0, 0xc420ed0820)
	/go2/src/github.com/cockroachdb/cockroach/pkg/server/server.go:687 +0x149
github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith.func1(0xc4200adb00, 0x2ab14ee3b640, 0xc420326630, 0xc42047a028, 0x2601ac0, 0xc420ed0820, 0xc4211780f0)
	/go2/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:142 +0xa9
created by github.com/cockroachdb/cockroach/pkg/util/netutil.(*Server).ServeWith
	/go2/src/github.com/cockroachdb/cockroach/pkg/util/netutil/net.go:144 +0x267
exit status 2
FAIL	github.com/cockroachdb/cockroach/pkg/sql/logictest	0.875s
Makefile:203: recipe for target 'test' failed
make: *** [test] Error 1
```

CC @knz

--- a/pkg/sql/parser/format.go
+++ b/pkg/sql/parser/format.go
@@ -108,6 +108,14 @@ func FmtReformatTableNames(
 	return &f
 }
 
+// StripTypeFormatting removes the flag that extracts types from the format flags,
+// so as to enable rendering expressions for which types have not been computed yet.
+func StripTypeFormatting(f FmtFlags) FmtFlags {
+	nf := *f
+	nf.showTypes = false
+	return &nf
+}
+
 // FmtExpr returns FmtFlags that indicate how the pretty-printer
 // should format expressions.
 func FmtExpr(base FmtFlags, showTypes bool, symbolicVars bool, showTableAliases bool) FmtFlags {

--- a/pkg/sql/subquery.go
+++ b/pkg/sql/subquery.go
@@ -73,13 +73,11 @@ func (s *subquery) Format(buf *bytes.Buffer, f parser.FmtFlags) {
 	if s.execMode == execModeExists {
 		buf.WriteString("EXISTS ")
 	}
-	if f == parser.FmtShowTypes {
-		// TODO(knz/nvanbenschoten): It is not possible to extract types
-		// from the subquery using Format, because type checking does not
-		// replace the sub-expressions of a SelectClause node in-place.
-		f = parser.FmtSimple
-	}
-	parser.FormatNode(buf, f, s.subquery)
+	// Note: we remove the flag to print types, because subqueries can
+	// be printed in a context where type resolution has not occurred
+	// yet. We do not use FmtSimple directly however, in case the
+	// caller wants to use other flags (e.g. to anonymize identifiers).
+	parser.FormatNode(buf, parser.StripTypeFormatting(f), s.subquery)
 }
 
 func (s *subquery) String() string { return parser.AsString(s) }

